### PR TITLE
Feature: add folowup functionality

### PR DESCRIPTION
**Pull request by muizkhan27**

### CODE REVIEW

File: followup.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+                except Exception as e:
+                    print(contact_email + ' error adding initial message')
+                    print(e)
+                
+                #format and prepend systemprompt
+                llm_sorted_messages.insert(0, {'role': 'system', 'content': formatted_system_prompt})
+                secret_message = '''[secret internal message] Hi Taylor, this is Heather, the CEO of Reposite. The lead 
+                you've been emailing hasn't responded in 3 days. Reply to this message with a followup email to the 
+                lead. Remember, the lead doesn't see this message. Do not acknowledge it in your response.'''
+                llm_sorted_messages.append({'role': 'user', 'content': secret_message})
```
Comment (uqarni) [inline followup.py:178]: let's log llm_sorted_messages somewhere for QC for a week or so

File: followup.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                # #generate response
+
+                for i in range(5):
+                    try:
+                        response = openai.ChatCompletion.create(model="gpt-4", temperature = 0, messages=llm_sorted_messages, max_tokens=600, user='muiz')
+                        usage = response['usage']
```
Comment (uqarni) [inline followup.py:184]: we need to add usage into the tokens columns of messages when we append the outbound message to our DB

File: followup_response.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        about this.
+        Should you change your mind and decide to explore building a more consistent pipeline of
+        group business, feel free to reach out or check out the membership benefits
+        here: https://www.reposite.io/membership-overview-1. Have a great day!
+        Lead: I said to stop messaging me
+        Assistant: I apologize if my messages have been intrusive, uzair. I respect your request and
+        will refrain from sending further communications. Should you have any change of mind about
+        our offerings at Reposite, you can always check out the membership details
+        here: https://www.reposite.io/membership-overview-1. Wishing you the best.
+        output = no
```
Comment (uqarni) [inline followup_response.py:62]: lol. ok we should improve these a bit to be more realistic.

File: followup_response.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        My sincerest apologies for any inconvenience caused.
+        Lead: I said stop
+        Lead: Actually, I would like to know more. Can I stay on the free tier or do I have to pay the
+        membership?
+        Assistant: Absolutely, Uzair! You can certainly stay on the Reposite Basic plan. However, I would
+        like to draw attention to the fact that with this plan, you will only be able to receive ad hoc
+        direct inbound requests and won't have a convenient level of transparency or abilities to
+        promote
+        your services.
+        output = yes
```
Comment (uqarni) [inline followup_response.py:77]: hahaha. let's improve these.

File: followup_response.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            model="gpt-3.5-turbo-16k",
+            temperature = 0,
+            messages=all_messages,
+            functions=functions,
+            function_call= {"name": "followup"},
+            user = 'muiz'
+        )
+        if response:
+            message = response["choices"][0]["message"]
+            error = ''
```
Comment (uqarni) [inline followup_response.py:100]: we actually need also track usage here as well. let's discuss.

File: llmwrapper.py [hunk 6]
```diff
@@ -120,7 +120,7 @@ def summarizer(messages, lead_email, org_email):
     to_summarize.insert(0, system_prompt)
     for i in range(5):
         try:
-            response = openai.ChatCompletion.create(model="gpt-4", messages=to_summarize, max_tokens=200)
+            response = openai.ChatCompletion.create(model="gpt-4", temperature = 0, messages=to_summarize, max_tokens=200, user='muiz')
```
Comment (uqarni) [inline llmwrapper.py:123]: user = 'reposite'

File: worker.py [hunk 7]
```diff
@@ -319,13 +322,13 @@ def main():
...
+        channel_id = os.environ.get("CHANNEL_ID")
+        channel_account_id = os.environ.get("CHANNEL_ACCOUNT_ID")
         recipient_actor_id = 'V-' + id
         recipient_email = crm.get_actor_info(recipient_actor_id)['email']
-        sender_actor_id = 'A-60198305'
-        sender_email = 'tbowdin@reposite.io'
-        subject = 'Re: Welcome to Reposite'
+        sender_actor_id = os.environ.get("SENDER_ACTOR_ID")
+        sender_email = os.environ.get("SENDER_EMAIL")
+        subject = os.environ.get("SUBJECT")
```
Comment (uqarni) [inline worker.py:332]: thanks for all this. appreciated.

File: worker.py [hunk 8]
```diff
@@ -381,14 +384,12 @@ def main():
...
+    # logging.basicConfig(level=logging.INFO)
+    # logging.info("Starting the scheduler. Will run every 15 minutes.")
     
-    scheduler.start()
-
-
-
-
+    # scheduler.start()
+    call_followup()
```
Comment (uqarni) [inline worker.py:395]: im confused here. we need to keep the 15-minute responder going. the followup will be deployed separately. it should run once a day.

**Review by uqarni — CHANGES_REQUESTED** [review]
looking good, let's address my comments together tomorrow.

File: followup.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    # Calculate three business days ago
+    followup_threshold_count = datetime.now()
+    while business_days_count < int(threshold_days):
+        followup_threshold_count -= timedelta(days=1)
+        if not is_weekend(followup_threshold_count):
+            business_days_count += 1
+    return followup_threshold_count
+
+
+def followup(db):
```
[the remark this answers is not in the document]
Comment (muizkhan27) [reply followup.py:45]: Yes I ll be doing it in the refactoring part that I am focusing on Today

File: followup.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+                membership_link = os.environ.get("MEMBERSHIP_LINK")
+                details = crm.get_contact_info(contact['id'], [
+                    'email', 'supplierorganizationname', 'firstname', 'lastname', 'prior_nmqrs',
+                    'quote_lead_company_name', 'quote_lead_category', 'lastquoterequestreceiveddate', 'quote_lead_destination', 
+                    'quote_lead_group_size', 'quote_lead_start_date', 'quote_lead_end_date', 'nmqrurl','ntmcampaignbucket'
+                ])
+                details = details['properties']
+                #gotta use this variable to pull initial message and select system prompt
+                ntmcampaignbucket = details.get('ntmcampaignbucket', "unknown")
+                if ntmcampaignbucket == None:
```
[the remark this answers is not in the document]
Comment (muizkhan27) [reply followup.py:84]: Yes we can

File: followup.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+                    'category': category,
+                    'date': date,
+                    'current_date': current_date,
+                    'destination': destination,
+                    'group_size': group_size,
+                    'trip_dates': trip_dates,
+                }
+                # print("\n\n These are the prompt variable for ", contact['contact_email'], ':\n ', prompt_variables, "\n\n")
+                examples = ""
+                temp_formatted_system_prompt = system_prompt.format(**prompt_variables)
```
[the remark this answers is not in the document]
Comment (muizkhan27) [reply followup.py:148]: It was with example as I was testing it and it ll be dry after refactoring

File: followup.py (hunk 1 above)
Comment (muizkhan27) [reply followup.py:178]: Sure, Will be done.

File: followup.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+                llm_sorted_messages.insert(0, {'role': 'system', 'content': formatted_system_prompt})
+                secret_message = '''[secret internal message] Hi Taylor, this is Heather, the CEO of Reposite. The lead 
+                you've been emailing hasn't responded in 3 days. Reply to this message with a followup email to the 
+                lead. Remember, the lead doesn't see this message. Do not acknowledge it in your response.'''
+                llm_sorted_messages.append({'role': 'user', 'content': secret_message})
+                # #generate response
+
+                for i in range(5):
+                    try:
+                        response = openai.ChatCompletion.create(model="gpt-4", temperature = 0, messages=llm_sorted_messages, max_tokens=600, user='muiz')
```
[the remark this answers is not in the document]
Comment (muizkhan27) [reply followup.py:183]: Okay, Noted.

File: followup.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (muizkhan27) [reply followup.py:84]: Resolved

File: followup.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (muizkhan27) [reply followup.py:148]: Resolved

File: followup.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (muizkhan27) [reply followup.py:183]: Done

File: llmwrapper.py (hunk 6 above)
Comment (muizkhan27) [reply llmwrapper.py:123]: Resolved

File: worker.py (hunk 7 above)
Comment (muizkhan27) [reply worker.py:332]: Your welcome

File: worker.py (hunk 8 above)
Comment (muizkhan27) [reply worker.py:395]: I have add another job in the scheduler to run followup every at 6:15 AM.

File: worker.py (hunk 7 above)
Comment (muizkhan27) [reply worker.py:332]: Resolved

File: followup_response.py (hunk 5 above)
Comment (muizkhan27) [reply followup_response.py]: Its done

**Pull request merged**